calico_containers/pycalico/ipam.py

```diff
@@ -832,3 +1034,116 @@ def get_assignment_attributes(self, address):
+            try:
+                self._claim_block_affinity(host, block_cidr, ipam_config)
+            except HostAffinityClaimedError:
+                unclaimed.append(cidr)
+                break
+            else:
+                claimed.append(cidr)
+
+        return claimed, unclaimed
+
+    @handle_errors
+    def release_affinity(self, cidr, host=None):
+        """
+        :param cidr: The CIDR covering the blocks to be released.  Raises a
+        InvalidBlockSizeError if the CIDR is smaller than the minimum allowable
+        block size.
+        :param host: (optional) The host ID to compare against the affinity of
+        each block that is being released.
+
+        :return: A tuple of:
+                 ([IPNetwork<blocks released>],
+                  [IPNetwork<blocks that were not claimed>],
+                  [IPNetwork<blocks that were claimed by another host>])
+        """
+        assert isinstance(cidr, IPNetwork)
+        if not validate_block_size(cidr):
+            _log.info("Requested CIDR %s is too small", cidr)
+            raise InvalidBlockSizeError("The requested CIDR is smaller than "
+                                        "the minimum block size.")
+        if not host:
+            host = get_hostname()
+
+        released = []
+        not_claimed = []
+        claimed_by_other = []
+
+        for block_cidr in cidr.subnet(BLOCK_PREFIXLEN[cidr.version]):
+            try:
+                self._release_block_affinity(host, block_cidr)
+            except HostAffinityClaimedByOtherError:
+                claimed_by_other.append(cidr)
+                break
+            except KeyError:
+                not_claimed.append(block_cidr)
+                break
+            else:
+                released.append(block_cidr)
+
+            return released, not_claimed, claimed_by_other
+
+    @handle_errors
+    def release_host_affinities(self, host):
+        """
+        Release host affinities, for all blocks owned by the host.
+
+        :param host: (optional) The host ID to compare against the affinity of
+        each block that is being released.
+        """
+        if not host:
+            host = get_hostname()
+
+        # Find all of the affine blocks that are listed for the host, and
+        # release affinity for each.  Note that the host may over-estimate
+        # which blocks it has affinity for so ignore any error indicating that
+        # the block is owned by another host - we simply won't release that
+        # block.
+        _log.debug("Releasing affinities for %s", host)
+        for version in (4, 6):
+            cidrs = self._get_affine_blocks(host, version, None)
+            for cidr in cidrs:
+                try:
+                    self._release_block_affinity(host, cidr)
+                except HostAffinityClaimedByOtherError:
+                    _log.info("Affine block %s is not owned by host %s - skip",
```

Comment:
Presumably this branch is to handle the race condition where we get the blocks for this host, another client releases one of them, and then we try to release it?

I'd add a comment explaining how we could hit this branch.